cirq/experiments/fidelity_estimation.py

```diff
@@ -372,3 +395,254 @@ def least_squares_xeb_fidelity_from_probabilities(
     return least_squares_xeb_fidelity_from_expectations(
         measured_expectations, exact_expectations, uniform_expectations
     )
+
+
+@dataclass(frozen=True)
+class _Sample2qXEBTask:
+    """Helper container for grouping a circuit to be sampled.
+
+    `prepared_circuit` is the full-length circuit (with index `circuit_i`. that
+    has been truncated to `cycle_depth` and has a measurement gate on it.
+    """
+
+    cycle_depth: int
+    circuit_i: int
+    prepared_circuit: 'cirq.Circuit'
+
+
+class _SampleInBatches:
+    def __init__(self, sampler: 'cirq.Sampler', repetitions: int):
+        """This closure will execute a list of `tasks` with one call to
+        `run_batch` on the provided sampler for a given number of repetitions."""
+        self.sampler = sampler
+        self.repetitions = repetitions
+
+    def __call__(self, tasks: List[_Sample2qXEBTask]):
+        prepared_circuits = [task.prepared_circuit for task in tasks]
+        results = self.sampler.run_batch(prepared_circuits, repetitions=self.repetitions)
+        assert len(results) == len(tasks)
+        records = []
+        for task, nested_result in zip(tasks, results):
+            (result,) = nested_result  # remove nesting due to potential sweeps.
+            sampled_inds = result.data.values[:, 0]
+            sampled_probs = np.bincount(sampled_inds, minlength=2 ** 2) / len(sampled_inds)
+
+            records += [
+                {
+                    'circuit_i': task.circuit_i,
+                    'cycle_depth': task.cycle_depth,
+                    'sampled_probs': sampled_probs,
+                }
+            ]
+        return records
+
+
+def _verify_and_get_two_qubits_from_circuits(circuits: Sequence['cirq.Circuit']):
+    """Make sure each of the provided circuits uses the same two qubits and return them."""
+    all_qubits_set: Set['cirq.Qid'] = set()
+    all_qubits_set = all_qubits_set.union(*(circuit.all_qubits() for circuit in circuits))
+    all_qubits_list = sorted(all_qubits_set)
+    if len(all_qubits_list) != 2:
+        raise ValueError(
+            "`circuits` should be a sequence of circuits each operating on the same two qubits."
+        )
+    return all_qubits_list
+
+
+class _NoProgress:
+    """Dummy (lack of) tqdm-style progress bar."""
+
+    def __init__(self, total: int):
+        pass
+
+    def __enter__(
+        self,
+    ):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def update(self, increment: int):
+        pass
+
+
+def sample_2q_xeb_circuits(
+    sampler: 'cirq.Sampler',
+    circuits: Sequence['cirq.Circuit'],
+    cycle_depths: Sequence[int],
+    repetitions: int = 10_000,
+    batch_size: int = 9,
+    progress_bar=tqdm.tqdm,
```

Comment:
Missing type information

Reply:
I'm loath to figure out what the annotation is for "thing that satisfies the context manager protocol"

Reply:
I can change it to `Optional[Type]` but that doesn't really clarify it

Reply:
`typing.ContextManager[Foo]`, where `Foo` the type returned by `__enter__` (or yielded if using `contextlib.contextmanager` decorator).

Reply:
`Optional[Callable[..., ContextManager]]` works, but again I'm not sure if that helps anything

Reply:
I don't know why `Optional[Callable[[int], ContextManager]]` doesn't work:

```
cirq/experiments/fidelity_estimation.py:526: error: Unexpected keyword argument "total"
```

which is the line where I'm calling the callable with an `int` argument .....